dialogue-hc4-client/src/main/java/com/palantir/dialogue/hc4/ApacheHttpClientBlockingChannel.java

```diff
@@ -0,0 +1,199 @@
+import org.apache.http.client.methods.RequestBuilder;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicHeader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class ApacheHttpClientBlockingChannel implements BlockingChannel {
+    private static final Logger log = LoggerFactory.getLogger(ApacheHttpClientBlockingChannel.class);
+
+    private final CloseableHttpClient client;
+    private final UrlBuilder baseUrl;
+
+    ApacheHttpClientBlockingChannel(CloseableHttpClient client, URL baseUrl) {
+        this.client = client;
+        this.baseUrl = UrlBuilder.from(baseUrl);
+    }
+
+    @Override
+    public Response execute(Endpoint endpoint, Request request) throws IOException {
+        // Create base request given the URL
+        UrlBuilder url = baseUrl.newBuilder();
+        endpoint.renderPath(request.pathParams(), url);
+        request.queryParams().forEach(url::queryParam);
+        URL target = url.build();
+        RequestBuilder builder =
+                RequestBuilder.create(endpoint.httpMethod().name()).setUri(target.toString());
+
+        // Fill headers
+        request.headerParams().forEach(builder::addHeader);
+
+        if (request.body().isPresent()) {
+            Preconditions.checkArgument(
+                    endpoint.httpMethod() != HttpMethod.GET, "GET endpoints must not have a request body");
+            RequestBody body = request.body().get();
+            builder.setEntity(new RequestBodyEntity(body));
+        }
+        return new HttpClientResponse(client.execute(builder.build()));
+    }
+
+    private static final class HttpClientResponse implements Response {
+
+        private final CloseableHttpResponse response;
+        private Map<String, List<String>> headers;
+
+        HttpClientResponse(CloseableHttpResponse response) {
+            this.response = response;
+        }
+
+        @Override
+        public InputStream body() {
+            try {
+                return response.getEntity().getContent();
+            } catch (IOException e) {
+                throw new SafeRuntimeException("Failed to get response stream", e);
+            }
+        }
+
+        @Override
+        public int code() {
+            return response.getStatusLine().getStatusCode();
+        }
+
+        @Override
+        public Map<String, List<String>> headers() {
+            if (headers == null) {
+                Map<String, List<String>> tmpHeaders = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+                for (Header header : response.getAllHeaders()) {
+                    String value = header.getValue();
+                    if (value != null) {
+                        tmpHeaders
+                                .computeIfAbsent(header.getName(), _name -> new ArrayList<>(1))
+                                .add(header.getValue());
+                    }
+                }
+                headers = tmpHeaders;
+            }
```

Comment:
seeing the lazy initialization of this makes me wonder if we should document the thread-safety expectations of Response somewhere?